File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/createvote/VHCreateOptionItem.java
package com.heaton.funnyvote.ui.createvote;

import androidx.recyclerview.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;

import com.heaton.funnyvote.database.Option;
import com.heaton.funnyvote.databinding.CardViewCreateVoteOptionBinding;

/**
 * Created by heaton on 2016/9/2.
 */

public class VHCreateOptionItem extends RecyclerView.ViewHolder {

    private final CardViewCreateVoteOptionBinding binding;
    private optionEditTextListener optionEditTextListener;
    private Option option;
    private CreateVoteTabOptionFragment.OptionItemListener itemListener;

    public VHCreateOptionItem(View itemView, CreateVoteTabOptionFragment.OptionItemListener itemListener) {
        this(CardViewCreateVoteOptionBinding.bind(itemView), itemListener);
    }

    public VHCreateOptionItem(CardViewCreateVoteOptionBinding binding, CreateVoteTabOptionFragment.OptionItemListener itemListener) {
        super(binding.getRoot());
        this.binding = binding;
        this.itemListener = itemListener;

        binding.relAdd.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                addNewOption();
            }
        });
        binding.imgDeleteOption.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                removeOption();
            }
        });
    }

    public void setLayout(int viewType, Option option) {
        this.option = option;
        if (viewType == OptionCreateItemAdapter.VIEW_TYPE_ADD_OPTION) {
            binding.relNormal.setVisibility(View.INVISIBLE);
            binding.relAdd.setVisibility(View.VISIBLE);
            binding.imgDeleteOption.setVisibility(View.GONE);
            binding.edtOptionTitle.setVisibility(View.GONE);
            binding.edtOptionTitle.removeTextChangedListener(optionEditTextListener);
        } else if (viewType == OptionCreateItemAdapter.VIEW_TYPE_NORMAL_OPTION) {
            binding.relNormal.setVisibility(View.VISIBLE);
            binding.relAdd.setVisibility(View.INVISIBLE);
            binding.imgDeleteOption.setVisibility(View.VISIBLE);
            binding.txtOptionNumber.setText(Integer.toString(getAdapterPosition() + 1));
            binding.edtOptionTitle.setVisibility(View.VISIBLE);
            binding.edtOptionTitle.removeTextChangedListener(optionEditTextListener);
            binding.edtOptionTitle.setText(option.getTitle());
            if (optionEditTextListener == null) {
                optionEditTextListener = new optionEditTextListener(itemListener);
            }
            binding.edtOptionTitle.addTextChangedListener(optionEditTextListener);
        }
    }

    public void addNewOption() {
        itemListener.onOptionAddNew();
    }

    public void removeOption() {
        itemListener.onOptionRemove(option.getId());
    }

    private final class optionEditTextListener implements TextWatcher {

        CreateVoteTabOptionFragment.OptionItemListener itemListener;
        public optionEditTextListener(CreateVoteTabOptionFragment.OptionItemListener itemListener) {
            this.itemListener = itemListener;
        }
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            itemListener.onOptionTextChange(option.getId(),s.toString());
        }

        @Override
        public void afterTextChanged(Editable s) {

        }
    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/createvote/VHCreateOptionItem.kt
package com.heaton.funnyvote.ui.createvote

import android.text.Editable
import android.text.TextWatcher
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.heaton.funnyvote.database.Option
import com.heaton.funnyvote.databinding.CardViewCreateVoteOptionBinding

class VHCreateOptionItem(
    val binding: CardViewCreateVoteOptionBinding,
    private val itemListener: CreateVoteTabOptionFragment.OptionItemListener?
) : RecyclerView.ViewHolder(binding.root) {

    constructor(
        itemView: View,
        itemListener: CreateVoteTabOptionFragment.OptionItemListener?
    ) : this(CardViewCreateVoteOptionBinding.bind(itemView), itemListener)

    private var optionEditTextListener: OptionEditTextListener? = null
    private lateinit var option: Option

    fun setLayout(viewType: Int, option: Option) {
        this.option = option
        if (viewType == OptionCreateItemAdapter.VIEW_TYPE_ADD_OPTION) {
            binding.relNormal.visibility = View.INVISIBLE
            binding.relAdd.visibility = View.VISIBLE
            binding.imgDeleteOption.visibility = View.GONE
            binding.edtOptionTitle.visibility = View.GONE
            binding.edtOptionTitle.removeTextChangedListener(optionEditTextListener)
        } else if (viewType == OptionCreateItemAdapter.VIEW_TYPE_NORMAL_OPTION) {
            binding.relNormal.visibility = View.VISIBLE
            binding.relAdd.visibility = View.INVISIBLE
            binding.imgDeleteOption.visibility = View.VISIBLE
            binding.txtOptionNumber.text = (adapterPosition + 1).toString()
            binding.edtOptionTitle.visibility = View.VISIBLE
            binding.edtOptionTitle.removeTextChangedListener(optionEditTextListener)
            binding.edtOptionTitle.setText(option.title)
            if (optionEditTextListener == null) {
                optionEditTextListener = OptionEditTextListener(itemListener)
            }
            binding.edtOptionTitle.addTextChangedListener(optionEditTextListener)
        }
        binding.relAdd.setOnClickListener { itemListener?.onOptionAddNew() }
        binding.imgDeleteOption.setOnClickListener { itemListener?.onOptionRemove(option.id) }
    }

    private inner class OptionEditTextListener(
        internal var itemListener: CreateVoteTabOptionFragment.OptionItemListener?
    ) : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            itemListener?.onOptionTextChange(option.id, s.toString())
        }
        override fun afterTextChanged(s: Editable) {}
    }
}
